### PR TITLE
[dv,top-level,escalation] Add two more escalation tests

### PR DESCRIPTION
**Pull request by matutem**
Add tests for TopEarlgreyAlertIdRvPlicFatalFault and TopEarlgreyAlertIdLcCtrlFatalStateError fatal errors.

Addresses two cases from #15649

Signed-off-by: Guillermo Maturana <maturana@google.com>

### CODE REVIEW

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv [hunk 1]
```diff
@@ -33,8 +33,7 @@ class chip_sw_all_escalation_resets_vseq extends chip_sw_base_vseq;
     '{"*keymgr*prim_reg_we_check*", TopEarlgreyAlertIdKeymgrFatalFaultErr},
     '{"*kmac*prim_reg_we_check*", TopEarlgreyAlertIdKmacFatalFaultErr},
     // TODO TopEarlgreyAlertIdLcCtrlFatalProgError: done in sw/device/tests/sim_dv/lc_ctrl_program_error.c?
-    // TODO TopEarlgreyAlertIdLcCtrlFatalStateError. These don't have onehot_checkers, so they need an
-    // alternative mechanism: maybe need to cause both u_reg, and u_reg_tap checkers?
+    '{"*lc_ctrl*state_regs*", TopEarlgreyAlertIdLcCtrlFatalStateError},
```
Comment (matutem) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:36]: Notice the `state_regs` string implies using sparse_fsm fault injection.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv [hunk 2]
```diff
@@ -47,11 +46,11 @@ class chip_sw_all_escalation_resets_vseq extends chip_sw_base_vseq;
     '{"*pwrmgr*prim_reg_we_check*", TopEarlgreyAlertIdPwrmgrAonFatalFault},
     '{"*rom_ctrl*prim_reg_we_check*", TopEarlgreyAlertIdRomCtrlFatal},
     '{"*rstmgr*prim_reg_we_check*", TopEarlgreyAlertIdRstmgrAonFatalFault},
-    // TODO kTopEarlgreyAlertIdRstmgrAonFatalCnstyFault
+    // TopEarlgreyAlertIdRstmgrAonFatalCnstyFault cannot use this test.
     // TODO TopEarlgreyAlertIdRvCoreIbexFatalSwErr
     '{"*rv_core_ibex*prim_reg_we_check*", TopEarlgreyAlertIdRvCoreIbexFatalHwErr},
     '{"*rv_dm*prim_reg_we_check*", TopEarlgreyAlertIdRvDmFatalFault},
-    // TODO kTopEarlgreyAlertIdRvPlicFatalFault
+    '{"*rv_plic*prim_reg_we_check*", TopEarlgreyAlertIdRvPlicFatalFault},
```
Comment (matutem) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:53]: This was a simple missed test.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv [hunk 3]
```diff
@@ -70,8 +69,8 @@ class chip_sw_all_escalation_resets_vseq extends chip_sw_base_vseq;
   rand int ip_index;
   constraint ip_index_c {ip_index inside {[0 : ip_alerts.size() - 1]};}
 
-  function int get_ip_index_from_name(string ip_name);
-    string regex = {"*", ip_name, "*prim_reg_we_check*"};
+  function int get_ip_index_from_name(string ip_name, string sec_cm);
```
Comment (matutem) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:72]: The reset of the changes in this file are to handle sparse fsm faults, since the sec_cm type is different.

File: sw/device/lib/testing/alert_handler_testutils.c [hunk 4]
```diff
@@ -69,19 +69,17 @@ void alert_info_to_string(const alert_info_t *info) {
            info->class_accum_cnt[0], info->class_accum_cnt[1],
            info->class_accum_cnt[2], info->class_accum_cnt[3]);
   LOG_INFO("loc_alert_cause=0x%x", info->loc_alert_cause);
-  char cause_bits[81];
-  cause_bits[80] = '\0';
-  int string_index = 0;
+  int set_count = 0;
```
Comment (matutem) [inline sw/device/lib/testing/alert_handler_testutils.c:72]: The previous way to show the alert_cause bits was pretty bad :(
Showing the on bits only is much better, since most of the time very few bits are on.

File: sw/device/tests/sim_dv/all_escalation_resets_test.c [hunk 5]
```diff
@@ -1003,8 +1008,9 @@ void check_alert_dump() {
     if (i == kExpectedAlertNumber) {
       CHECK(actual_info.alert_cause[i], "Expected alert cause %d to be set", i);
     } else {
-      CHECK(!actual_info.alert_cause[i], "Expected alert cause %d to be clear",
-            i);
+      if (actual_info.alert_cause[i]) {
```
Comment (matutem) [inline sw/device/tests/sim_dv/all_escalation_resets_test.c:1011]: There are cases where a fatal alert generates another one, for example, some lc_ctrl faults lead to otp_ctrl faults, so this relaxes the requirement.

File: hw/ip/flash_ctrl/rtl/flash_ctrl.sv [hunk 6]
```diff
@@ -331,7 +331,7 @@ module flash_ctrl
   // flash disable declaration
   mubi4_t [FlashDisableLast-1:0] flash_disable;
 
-  // flash control arbitration between softawre / harware interfaces
+  // flash control arbitration between software and hardware interfaces
```
Comment (marnovandermaas) [inline hw/ip/flash_ctrl/rtl/flash_ctrl.sv:334]: nit: maybe put this in a separate commit.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv [hunk 7]
```diff
@@ -23,7 +23,7 @@ class chip_sw_all_escalation_resets_vseq extends chip_sw_base_vseq;
     '{"*edn1*prim_reg_we_check*", TopEarlgreyAlertIdEdn1FatalAlert},
     '{"*entropy_src*prim_reg_we_check*", TopEarlgreyAlertIdEntropySrcFatalAlert},
     '{"*flash_ctrl*prim_reg_we_check*", TopEarlgreyAlertIdFlashCtrlFatalStdErr},
-    // TODO test u_eflash.u_flash alert TopEarlgreyAlertIdFlashCtrlFatalErr and
+    // TODO test u_eflash.u_flash alert TopEarlgreyAlertIdFlashCtrlFatalErr
```
Comment (marnovandermaas) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:26]: Isn't this and connecting the two TODO lines?

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv [hunk 8]
```diff
@@ -47,11 +46,11 @@ class chip_sw_all_escalation_resets_vseq extends chip_sw_base_vseq;
     '{"*pwrmgr*prim_reg_we_check*", TopEarlgreyAlertIdPwrmgrAonFatalFault},
     '{"*rom_ctrl*prim_reg_we_check*", TopEarlgreyAlertIdRomCtrlFatal},
     '{"*rstmgr*prim_reg_we_check*", TopEarlgreyAlertIdRstmgrAonFatalFault},
-    // TODO kTopEarlgreyAlertIdRstmgrAonFatalCnstyFault
+    // TopEarlgreyAlertIdRstmgrAonFatalCnstyFault cannot use this test.
     // TODO TopEarlgreyAlertIdRvCoreIbexFatalSwErr
     '{"*rv_core_ibex*prim_reg_we_check*", TopEarlgreyAlertIdRvCoreIbexFatalHwErr},
     '{"*rv_dm*prim_reg_we_check*", TopEarlgreyAlertIdRvDmFatalFault},
-    // TODO kTopEarlgreyAlertIdRvPlicFatalFault
```
Comment (marnovandermaas) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:54]: Why did you remove this one? Is is the same reason as removing `kTopEarlgreyAlertIdRstmgrAonFatalCnstyFault`? If so, maybe add a comment like you did for that one?

File: sw/device/lib/testing/alert_handler_testutils.c (hunk 4 above)
Comment (marnovandermaas) [reply sw/device/lib/testing/alert_handler_testutils.c:72]: Nit: maybe useful to add this comment directly into the source code.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv (hunk 3 above)
Comment (marnovandermaas) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:72]: Nit: maybe it is useful to add this comment into the source code?

File: sw/device/tests/sim_dv/all_escalation_resets_test.c [hunk 9]
```diff
@@ -323,11 +324,12 @@ static void lc_ctrl_fault_checker(bool enable, const char *ip_inst,
...
+  uint32_t bus_integ_error =
+      bitfield_bit32_write(0, kDifLcCtrlStatusCodeBusIntegError, true);
+  uint32_t state_error =
+      bitfield_bit32_write(0, kDifLcCtrlStatusCodeCorrupt, true);
   uint32_t expected_status =
-      enable
-          ? (bitfield_bit32_write(0, kDifLcCtrlStatusCodeBusIntegError, true) &
-             mask)
-          : 0;
+      enable ? (type == we_check ? bus_integ_error : state_error) : 0;
```
Comment (marnovandermaas) [inline sw/device/tests/sim_dv/all_escalation_resets_test.c:332]: This looks like an important fix. Does it make sense to put it in a separate commit with its own commit message?

File: sw/device/tests/sim_dv/all_escalation_resets_test.c (hunk 5 above)
Comment (marnovandermaas) [reply sw/device/tests/sim_dv/all_escalation_resets_test.c:1011]: Nit: maybe put this comment into the source code?

**Review by marnovandermaas — COMMENTED** [review]
Nice work! Just a few suggestions and nits below.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv (hunk 7 above)
Comment (matutem) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:26]: Yes, and both are TODO, each on its own merits :) They just happen to be in consecutive lines.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv (hunk 8 above)
Comment (matutem) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:54]: I removed it because this alert is not a TODO anymore: the replacement text means it is handled now.

File: sw/device/tests/sim_dv/all_escalation_resets_test.c (hunk 9 above)
Comment (matutem) [reply sw/device/tests/sim_dv/all_escalation_resets_test.c:332]: Well, this is not really a separate fix, but it is required to handle one of the two added alerts.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv (hunk 3 above)
Comment (matutem) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:72]: I added a comment to this function.

File: sw/device/lib/testing/alert_handler_testutils.c (hunk 4 above)
Comment (matutem) [reply sw/device/lib/testing/alert_handler_testutils.c:72]: Good idea. Done.

File: hw/ip/flash_ctrl/rtl/flash_ctrl.sv (hunk 6 above)
Comment (matutem) [reply hw/ip/flash_ctrl/rtl/flash_ctrl.sv:334]: I think it is okay to fold trivial changes with others: this change for example would never cause a test failure, so bisection is not an issue.

File: sw/device/tests/sim_dv/all_escalation_resets_test.c (hunk 5 above)
Comment (matutem) [reply sw/device/tests/sim_dv/all_escalation_resets_test.c:1014]: Done.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv (hunk 7 above)
Comment (marnovandermaas) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:26]: Ok thanks

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv (hunk 8 above)
Comment (marnovandermaas) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:54]: Ok thanks.

File: hw/ip/flash_ctrl/rtl/flash_ctrl.sv (hunk 6 above)
Comment (marnovandermaas) [reply hw/ip/flash_ctrl/rtl/flash_ctrl.sv:334]: Ok, yes I guess changing comments is relatively safe.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv (hunk 3 above)
Comment (marnovandermaas) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv:74]: Thanks!

File: sw/device/lib/testing/alert_handler_testutils.c (hunk 4 above)
Comment (marnovandermaas) [reply sw/device/lib/testing/alert_handler_testutils.c:72]: Thanks!

**Review by marnovandermaas — APPROVED** [review]
LGTM, although it would be good for someone else to have a look at it as well.

File: sw/device/tests/sim_dv/all_escalation_resets_test.c [hunk 10]
```diff
@@ -812,9 +815,12 @@ static void execute_test(const dif_aon_timer_t *aon_timer) {
       fault_checker = fc;
     } break;
     // TODO add mechanism to inject kTopEarlgreyAlertIdLcCtrlFatalProgError by
-    // forcing otp_prog_err_o from lc_ctrl_fsm and
-    // kTopEarlgreyAlertIdLcCtrlFatalStateError using sparse fsm.
-    // alerts, and corresponding CSR bit to check.
+    // forcing otp_prog_err_o from lc_ctrl_fsm
+    case kTopEarlgreyAlertIdLcCtrlFatalStateError: {
```
Comment (msfschaffner) [inline sw/device/tests/sim_dv/all_escalation_resets_test.c:819]: I am not sure this can even be checked in SW, since forcing the LC_CTRL FSM into a fatal error state will disable the CPU immediately (it behaves similarly to escalation or SCRAP).

**Review by msfschaffner — APPROVED** [review]
Thanks @matutem. One comment RE the LC_CTRL fatal fault, otherwise LGTM.

You may also have to rerun `make -k -C hw` (CI complains...).
